// lp-messaging-sdk/lib/profile/client-properties.js

// ClientProperties class based on UMS ClientProperties
// https://lpgithub.dev.lprnd.net/Architecture/api-ums/blob/master/api-ums-types/src/main/java/com/liveperson/api/ams/headers/ClientProperties.java

/**
 * ClientProperties describes the attributes and capabilities of a consumer's client
 * @category Connection
 */
class ClientProperties {
    constructor(a) {
        this.appId = a.appId || null;
        this.appVersion = a.appVersion || null;
        this.ipAddress = a.ipAddress || null;
        this.osVersion = a.osVersion || null;
        this.osName = a.osName || null;
        this.integrationVersion = a.integrationVersion || null;
        this.browser = a.browser || null;
        this.browserVersion = a.browserVersion || null;
        this.deviceManufacturer = a.deviceManufacturer || null;
        this.deviceModel = a.deviceModel || null;
        this.timeZone = a.timeZone || null;
        this.deviceFamily = a.deviceFamily || null;
        this.features = a.features || null;
        this.os = a.os || null;
        this.integration = a.integration || null;
    }

    /**
     * @param {Object} args
     * @param {String} args.appId - Application identifier
     * @param {String} args.appVersion - Application version identifier
     * @param {String} args.browser - Browser name
     * @param {String} args.browserVersion - Browser version
     * @param {ClientProperties.DeviceFamily} args.deviceFamily - Device family (e.g. "DESKTOP", "MOBILE", etc)
     * @param {String} args.deviceManufacturer - Device manufacturer (e.g. 'LG', 'Apple', etc)
     * @param {String} args.deviceModel - Device model
     * @param {Array.<ClientProperties.Features>} args.features - Array of possible client features
     * @param {ClientProperties.Integration} args.integration - Integration type.  For example: web_sdk, mobile_sdk, and brand_sdk
     * @param {String} args.integrationVersion - Identifies the version of the integration
     * @param {String} args.ipAddress - client IP address
     * @param {ClientProperties.OS} args.os - Operating system (e.g. 'WINDOWS', 'LINUX')
     * @param {String} args.osName - Text describing the operating system version
     * @param {String} args.osVersion - Text describing the operating system version
     * @param {String} args.timeZone - TimeZone identifier.  Input format is either TZ identifier (i.e. 'America/Los_Angeles' OR GMT Offset (i.e 'GMT-7:00' or 'GMT+5:30')
     * @returns {ClientProperties}
     */
    static createClientProperties(
        {
            appId = null,
            appVersion = null,
            browser = null,
            browserVersion = null,
            deviceFamily = null,
            deviceManufacturer = null,
            deviceModel = null,
            features = null,
            integration = null,
            integrationVersion = null,
            ipAddress = null,
            os = null,
            osName= null,
            osVersion = null,
            timeZone = null,
        }) {
        return new ClientProperties(arguments[0]);
    }
}

module.exports = ClientProperties
