// lp-messaging-sdk/lib/connection/connection-state.js
/**
 * typedef {string} ConnectionState
 */

/**
 * An enum of the possible connection states
 * @enum {string}
 * @category Connection
 * @example
 * if (connection.state === ConnectionState.CONNECTED) {
 *     console.log('connection is connected');
 * }
 */
const ConnectionState = {

    /** the initial state, no connection has been attempted */
    INITIAL: 'initial',

    /** the connection is in the process of connecting */
    CONNECTING: 'connecting',

    /** the connection is established and ready to be used */
    CONNECTED: 'connected',

    /** the connection has never been established */
    FAILED_TO_CONNECT: 'failed_to_connect',

    /** indicates that connection was connected but has disconnected */
    DISCONNECTED: 'disconnected',

    /** the connection is in the process of closing */
    CLOSING: 'closing',

    /** the connection has closed and was intentional */
    CLOSED: 'closed',

    /** no connection is possible or will be attempted */
    NULL: 'null'

};

module.exports = ConnectionState;
