// lp-messaging-sdk/lib/connection/connection-websocket-consumer.js
const WebsocketConnection = require('./connection-websocket');
const CSDS = require("../services/csds-cache");
const IDP = require('../services/idp');
const apiV3 = require('../api/ws').v3;
const {waitForEvent} = require("../common");

const consumerDefaultSubscriptionQuery = {
    stage:["OPEN", "CLOSE"]
};

const waitForMSNotificationTimeout = 1000 * 60 * 10; // 10 minutes

// todo come up with a better way of handling this
/*const defaultClientProperties = {
    // "appId": "com.tmo.lpapp",
    // "appVersion": "1.0.0.0",
    // "ipAddress": "127.0.0.1",
    // "deviceFamily": "DESKTOP",
    // "os": "LINUX",
    // "osVersion": "14.10",
    // "osName": "osName",
    // "integration": "WEB_SDK",
    // "integrationVersion": "11.00",
    // "browser": "CHROME",
    // "browserVersion": "47.0",
    // "timeZone": "Europe/Athens",
    // "deviceManufacture": "LG",
    // "deviceModel": "G5",
    "features": [
        //"AUTO_MESSAGES",
        //"CO_BROWSE",
        //"PHOTO_SHARING",
        //"FILE_SHARING",
        //"MARKDOWN_HYPERLINKS"
        "RICH_CONTENT",
        "QUICK_REPLIES",
        "MULTI_DIALOG"
    ]
};*/

/**
 * Represents a consumer side websocket connection to UMS.
 * @category Connection
 */
class ConsumerWebsocketConnection extends WebsocketConnection {

    constructor(a) {
        super(a);

        this.unauthConnectorId = a.unauthConnectorId || null;

        // allow the user to use the mock auth user flow
        this.useMockAuth = false; // default to false
        if (a.authData && a.authData.useMockAuth) {
            this.useMockAuth = true;
            this.customerName = a.authData.customerName || 'default';
        }

        this.token = a.token || null;

        // construct the client properties
        this.clientProperties = {};

        // add client properties
        if (a.clientProperties) {
            this.clientProperties = a.clientProperties;
        }

        // the query used when making default subscription
        // TODO what if they passed in invalid query?
        // UMS seems to allow random strings
        this.defaultSubscriptionQuery = a.defaultSubscriptionQuery === null ?
            consumerDefaultSubscriptionQuery :
            a.defaultSubscriptionQuery;
    }

    get api() {
        return apiV3;
    }

    /**
     * Returns the token from this connection
     * @returns {Promise<string|null>} JWT token
     */
    async getToken() {
        return this.token;
    }

    async _onConnect() {
        // do the normal onConnect stuff
        await super._onConnect();

        this.userProfile = await this._getUserProfile();
        this.userId = this.userProfile.userId;
    }

    /**
     * Consumer specific authVerify call
     * @private
     */
    async _verifyRequest() {
        // send the authentication message (required for v3)
        const initRequest = this.api.initConnection(this.token, this.clientProperties);
        await this.send(initRequest);
    }

    /**
     * for consumer flow, requires InitConnection to provide auth token
     * @param accountId
     * @param protocol
     * @returns {Promise<string>}
     * @private
     */
    static async _getWebsocketUrlConsumer(accountId, appId, appVersion) {
        const host = await CSDS.lookup(accountId, CSDS.ServiceName.asyncMessagingEnt);
        const userType = 'consumer';
        const trackingParams = WebsocketConnection.createTrackingUrlParamString(appId, appVersion);
        return `${CSDS.defaultWebsocketProtocol}://${host}/ws_api/account/${accountId}/messaging/${userType}?v=3&${trackingParams}`;
    }

    async _getWebsocketUrl() {
        // If appId or appVersion is not provided on the connection, use the clientProfile values if possible
        let appId = this.appId;
        if (!this.appId && this.clientProperties) {
            appId = this.clientProperties.appId;
        }

        let appVersion = this.appVersion;
        if (!appVersion && this.clientProperties) {
            appVersion = this.clientProperties.appVersion;
        }

        return ConsumerWebsocketConnection._getWebsocketUrlConsumer(this.accountId, appId, appVersion);
    }

    /**
     * Clear the current auth token, forces a _generateToken() when ._connect() is called again
     * @private
     */
    _onAuthInvalidError() {
        // TODO CO-6087
        this.token = null;
    }

    /**
     * Called during _connect() to generate a JWT token, if it does not exist
     * @returns {Promise<void>}
     * @private
     */
    async _initToken() {
        if (this.token !== null) {
            return;
        }

        await this._generateToken();
    }

    async _generateToken() {
        // get the auth information
        let authInfo = {start:Date.now()};

        // make the auth call
        this.emit('getConsumerToken#request', authInfo);
        this.token = await this._getToken();
        authInfo.end = Date.now();
        this.emit('getConsumerToken#response', authInfo);
    }

    // gets overridden by authenticated to create authenticated token
    async _getToken() {

        // a mock auth jwt generator, only to be used for testing
        if (this.useMockAuth) {
            return this._getMockAuthToken()
        }

        // sometimes the unauth connector won't exist, use signup call
        if (this.unauthConnectorId === null) {
            return this._getTokenWithoutConnectorId();
        }

        return this._getTokenWithConnectorId(this.unauthConnectorId);
    }

    async _getTokenWithoutConnectorId() {
        const token = await IDP.getSignupToken(this.accountId);
        return token;
    }

    async _getTokenWithConnectorId(connectorId) {
        const token0 = await IDP.getAnonymousToken(this.accountId);
        const token1 = await IDP.authenticateAnonymousToken(this.accountId, connectorId, token0);
        return token1;
    }

    async _getMockAuthToken() {
        const token = await IDP.getMockAuthToken(this.accountId, this.customerName);
        return token;
    }

    async _getMockStepUpToken(authConnectorId) {
        const token = await IDP.getMockStepupToken(this.accountId, authConnectorId, this.token, this.customerName);
        return token;
    }

    /**
     * Trigger a mock step up on this connection.
     * Only available in test environments.
     * @param authConnectorId - the authenticated connectorId
     * @param customerName - the name to be included in the auth SDEs
     */
    async mockStepUp(authConnectorId, customerName) {

        this.customerName = customerName;

        // first, replace the token
        this.token = await this._getMockStepUpToken(authConnectorId);

        return new Promise((resolve, reject) => {

            // wait for the connect
            this.once('connect', async () => {

                // change participant on the conversation
                for (const conversation of this._conversations.values()) {
                    await conversation.consumerTakeover(this.userId);
                }

                resolve();
            });

            // cause a disconnect
            this._forceDisconnect();

            // reconnect if we are not maintaining automatically
            if (!this.maintainConnection) {
                this.once('close', async () => {
                    await this._connectWithRetry();
                    resolve();
                });
            }
        });

    }

    _onDisconnect() {
        super._onDisconnect();

        // loop through all dialogs and set _isSubscribedMSEvents to false
        this._resetDialogSubscribeStatus();
    }

    _resetDialogSubscribeStatus() {
        //const conversations = this._conversations.values();
        for (let conversation of this._conversations.values()) {
            if (conversation.openDialog === null) continue;
            conversation.openDialog._isSubscribedMSEvents = false;
        }
    }

    async _onReconnect() {
        // resubscribe to MS events on known conversations
        await this._resubscribeToMessagingEvents();

        await super._onReconnect();
    }

    /**
     * Calls _subscribeMessagingEvents on any openDialogs that are not known to be currently subscribed
     * @private
     */
    async _resubscribeToMessagingEvents() {
        for (let conversation of this._conversations.values()) {

            // only subscribe if there is an open dialog
            if (conversation.openDialog === null) continue;

            // no need to double subscribe
            if (conversation.openDialog._isSubscribedMSEvents) continue;

            await conversation.openDialog._subscribeMessagingEvents();
        }
    }

    /**
     * Makes a request to UMS for a new conversation.
     * Constructs a conversation object with the response information.
     *
     * @async
     * @param {object} [args]
     * @param {string} [args.skillId] - creates this conversation in the specified skillId
     * @param {object} [args.context] - this data will be passed back in the notifications as conversationDetails.context
     * @returns {Promise<Conversation>} - the conversation object
     */
    async createConversation({skillId, context, campaignInfo} = {}) {
        const conversation = await super.createConversation({skillId, context, campaignInfo});

        // automatically subscribe to MS events
        await conversation.openDialog._subscribeMessagingEvents(false);

        return conversation;
    }

    /**
     * Called just before a subscription emits a previously unseen conversation
     * @private
     */
    async _onConversation(conversation) {
        await super._onConversation();
        try {
            // create MS subscription
            await conversation.openDialog._subscribeMessagingEvents(true);
        } catch (err) {
            this._emitError(err);
        }
    }

    /**
     * Represents how consumers create conversation
     *
     * @param args
     * @param args.skillId
     * @param args.context
     * @returns {Promise<uuid>} conversationId of the created conversation
     * @protected
     */
    async _consumerRequestConversation({skillId, context, campaignInfo}) {
        // make the request
        const request = this.api.consumerRequestConversation({skillId, conversationContext: context, campaignInfo});
        const result = await this.send(request);

        // catch a special error
        if (typeof result === 'string') {
            throw Error(result);
        }

        // conversationId is the only piece of data returned from UMS
        return result.conversationId;
    }

    async _subscribeMessagingEvents(conversationId, dialogId, waitForMessages) {
        // make the request
        const request = this.api.subscribeMessagingEvents(conversationId, dialogId);
        const result = await this.send(request);

        if (waitForMessages) {
            // waiting for 10 minutes because this should always resolve, if we don't get the notification event there are bigger issues
            await waitForEvent(this, 'messageNotification', waitForMSNotificationTimeout);
        }
    }

    /**
     * Allows an update of a subset of user profile fields
     *
     * @param {object} data
     * @param {string} [data.firstName]
     * @param {string} [data.lastName]
     * @param {string} [data.nickName]
     * @returns {Promise<void>}
     */
    async setUserProfile(data) {
        const result = await this.send(this.api.setUserProfile(data));
        if (result !== "OK User Profile set successfully") {
            throw new Error(result);
        }
    }

    async _sendCsat(conversationId, csatRate, csatResolutionConfirmation, csatStatus) {
        const request = this.api.updateCsat(conversationId, csatRate, csatResolutionConfirmation, csatStatus);
        const result = await this.send(request);

        // check for success messages
        if (result === 'OK CSAT processed successfully') {
            return;
        }

        // otherwise, it was a failure
        throw Error(result);
    }

}

module.exports = ConsumerWebsocketConnection;
