// lp-messaging-sdk/lib/api/rest/index.js
const RestApiVersion = {
    V1: 'RESTv1'
};

const v1 = {
    version: RestApiVersion.V1,

    // general
    getDownloadUrl: require('./general/get-download-url'),
    getUploadUrl: require('./general/get-upload-url'),

    // conversation
    addParticipant: require('./conversation/add-participant'),
    closeDialog: require('./conversation/close-dialog'),
    getConversationById: require('./conversation/get-conversation-by-id'),
    // TODO add:
    //  getDialogById
    //  getSingleParticipant
    //  removeParticipant
    transferConversation: require('./conversation/transfer-conversation'),
    updateConversationStage: require('./conversation/update-conversation-stage'),
    updateTTR: require('./conversation/update-ttr'),
    updateManualETTR: require('./conversation/update-manualettr'),

    // profile TODO add

    // messaging
    publishMessage: require('./messaging/publish-message'),
    publishAcceptStatus: require('./messaging/publish-accept-status'),
    publishRichContent: require('./messaging/publish-rich-content'),
    setChatState: require('./messaging/set-chat-state'),
    getMessages: require('./messaging/get-messages'),

    // notifications TODO TBD
    notificationTypes: {
        ExConversationChangeNotification: '.ams.aam.ExConversationChangeNotification',
        MessagingEventNotification: '.ams.ms.OnlineEventDistribution'
    }
};

module.exports = {v1};
